src/compiler.jl

```diff
@@ -1223,168 +1215,211 @@ function nested_codegen!(mod::LLVM.Module, f, tt)
-    # pfor threads tape[i] = aug(func, dfunc)
-
-    tt = Tuple{}
-    # TODO with the MI itself, we should be able to hoist the thunk generation into
-    # the original compilation (rather than runtime JIT compiling)
-    forward = thunk(func, dfunc, Const, tt, Val(API.DEM_ForwardMode), width)
-
-    function fwd()
-        forward()
-    end
-
-    Base.Threads.threading_run(fwd)
+function runtime_pfor_fwd(func, ptr, dfunc, ::Type{ThunkTy})::Cvoid where ThunkTy
+    thunk = ThunkTy(func, ptr, dfunc)
+    Base.Threads.threading_run(thunk)
     return
 end
 
-function runtime_pfor_augfwd(func, dfunc, width)::Int64
-    # tape = vec{Any}(numthreads())
-    # pfor threads tape[i] = aug(func, dfunc)
-
-    tt = Tuple{}
-    # TODO with the MI itself, we should be able to hoist the thunk generation into
-    # the original compilation (rather than runtime JIT compiling)
-    forward, adjoint = thunk(func, dfunc, Const, tt, Val(API.DEM_ReverseModePrimal), width)
+function runtime_pfor_augfwd(func, ptr, dfunc, ::Type{ThunkTy})::Ptr{Ptr{Cvoid}} where ThunkTy
+    thunk = ThunkTy(func, ptr, dfunc)
+    @show "starting augfwd", thunk
+    flush(stdout)
+    tapes = Base.unsafe_convert(Ptr{Ptr{Cvoid}}, Libc.malloc(sizeof(Ptr{Cvoid})*Base.Threads.nthreads()))
 
-    tapes = Vector{Ptr{Cvoid}}(undef, Base.Threads.nthreads())
-    # tapes = unsafe_convert(Ptr{Ptr{Cvoid}}, Libc.malloc(sizeof(Ptr{Cvoid})*Base.Threads.nthreads))
     function fwd()
-        tapes[Base.Threads.threadid()] = forward()[1]
+        @show "called fwd at tid", Base.Threads.threadid()
+        flush(stdout)
+        ntape = thunk()[1]
+        @show ntape, Base.Threads.threadid(), tapes
+        flush(stdout)
+        unsafe_store!(tapes, ntape, Base.Threads.threadid())
     end
-    adj = () -> adjoint(tapes[Base.Threads.threadid()])
-    id = getid()
-    leaked_objs[id] = adj
     Base.Threads.threading_run(fwd)
-    # Note: `adj` is an immutable object, and thus has pass-by-value semantics
-    #       directly returning here allocates a heap object whose pointer is then leaked
-    #       onto the tape and GC, will kill it. Until we have proper tape+GC support,
-    #       we stash the object in a global dict, and return the id of the object
-    return id
-end
-
-function runtime_pfor_rev(id::Int64)
-    tape = leaked_objs[id]
-    delete!(leaked_objs, id)
-    Base.Threads.threading_run(tape)
+    return tapes
+end
+
+function runtime_pfor_rev(func, ptr, dfunc, ::Type{ThunkTy}, tapes::Ptr{Ptr{Cvoid}}) where ThunkTy
+    thunk = ThunkTy(func, ptr, dfunc)
+    @show "starting rev", thunk
+    flush(stdout)
+    function rev()
+        ntape = unsafe_load(tapes, Base.Threads.threadid())
+        @show "rev", ntape, Base.Threads.threadid(), tapes
+        flush(stdout)
+        @show InteractiveUtils.@code_llvm thunk(ntape)
+        flush(stdout)
+        thunk(ntape)
+    end
+    Base.Threads.threading_run(rev)
+    @show "post run", thunk
```

Comment:
Weirdly enough introducing this show/flush removes one of the gc segfaults, though the result remains incorrect -- as if the reverse thunk code was never run